### DISCUSSION TITLE
Add ability to center editor camera on scenes when running current scene

### DISCUSSION THREAD

**Discussion by alexgamadev** [godotengine/godot-proposals#8970] [Editor]
Sorry if this has been mentioned before but I didn't see anything here.

When running a scene, objects tend to be centered on the origin (0, 0) and in a lot of cases the scenes don't have cameras because they don't need them. The problem is that when the scene runs, the camera isn't centered on 0,0 meaning the nodes within the scene show at the top left of the screen, sometimes partially obscured. If the scene is being built around 0,0, it doesn't really make much sense for the camera to be positioned with the top left at 0,0 because you don't normally build scenes only down and to the right. It is possible to get around this in some slightly hacky ways as far as I'm aware (Like autloading a centered camera) but really this should be a simple inbuilt change since it's so fundamental.

Correct me if I'm wrong but a centered viewport camera would be significantly more useful to the majority of users than the current top-left anchored one but maybe there's some common use case I'm missing.

**Comment by Calinou** [comment]
Changing this default behavior would break compatibility with existing projects that don't use a Camera2D, such as GUI-only applications.

Consider using the editor camera replication feature that is available when running a project (click the camera icon at the top of the 2D/3D editor viewport).

**Comment by alexgamadev** [comment]
But could it not just be an option that was available in the project settings? Like a viewport camera anchor that defaults to top left but can be changed to center if wanted. That way it wouldn't break any compatibility. I didn't know about the editor camera replication feature though so thanks for that suggestion! It is very cool but it still has the same issue of not being able to be defaulted to on. 

It might not seem like the biggest deal in the world but from a useability standpoint, the ability to center the editor camera on the origin by default just seems like such a fundamental feature for a game focused engine. Not trying to be difficult, just wondering if there's any genuine reason why this can't/shouldn't be allowed as a simple project setting?

**Reply by Calinou** [reply]
A opt-in project setting might be acceptable, especially since it would benefit many [official demos](https://github.com/godotengine/godot-demo-projects/) which currently use a dummy Camera2D to support multiple resolutions. This Camera2D needs to have its offset adjusted depending on the project's base window size too.

**Reply by alexgamadev** [reply]
Yeah that would be perfect. It prevents any break in compatibility but makes it extremely easy to default to a system that works better for games than is currently possible. I'm hoping it's not too difficult either since the scene replication feature you mentioned already deals with changing the editor camera position?

**Reply by hsandt** [reply]
I was about to open a discussion for exactly the same use-case (preview 2D objects) and found this thread. I also thought about a few other solutions so I'll sum up the advantages and disadvantages of what was discussed here + a few other ideas.

**Project settings**

I also think a Project Settings would be good. It should not be hidden too much, and could be exposed further in the form of a toggle near the Play Scene buttons.

One limitation is that even if the Center Origin feature is only applied in scenes without a camera, it would still misbehave in games that don't use a camera and just place their elements following the origin as top-left corner convention (typically a one-screen game). This can be resolved by adding a Camera2D exactly where the default would be (top-left at origin), though.

**Project Camera Override**

The Project Camera Override, whose button is also exposed in the scene panel, is also great, since the user would be looking at the object when editing it (although not perfectly centered). But we need to make that button clickable _before_ running the scene to effectively allow making it default. Also note that it zooms in unexpected ways when using window resolution scaling (at least in my pixel art project it appears much bigger in the game window as in the editor scene tab).

Its state should probably be saved per scene rather than per project so we only enable it for "object" scenes.

In this case the user needs to remember to enable this for every object scene they want to preview.

**Meta scene property**

I also considered a meta scene property, a flag that would say if the scene is an "actual scene" vs an "object", so that the adequate default behavior could be deduced; but I'm not sure if it's worth it (compared to adding a dummy Camera2D with top-left at origin in scenes that were relying on the default camera, in one-screen games).

 It could also become the default in future versions of Godot, only affecting new projects. It would still only apply to scenes without an existing Camera2D node, so 
 
**Dedicated button to Run Current Scene with Camera Centered on Origin if scene doesn't contain a Camera**

A dedicated button, e.g. mapped to Ctrl+F6, would also allow the user to quickly preview an object without having to toggle anything. It would be smart enough not to override an existing Camera so you could almost literally replace standard F6 (Run Current Scene) with this new button - but keep Run Current Scene when you really have one of those non-object scenes without a camera because it was more convenient (e.g. menus, cinematics...).

**Bonus question for 3D**

I'm not familiar with 3D, but how should we tackle 3D scenes? Use a default camera looking at the origin from a certain angle? Should the angle be tunable in editor options, the same way as 2D zoom could be tuned? (if using Project Camera override, we at least know what angle to use)

In my case I only need this for 2D right now.

**Proposal**

So, should we open a proposal now for what seems the most convenient, probably either a Project settings (+ a visible toggle button if possible) or extending Project Camera Override to allow permanent toggle? Or do you need further discussion on how to make this work?

**Reply by aXu-AP** [reply]
> I'm not familiar with 3D, but how should we tackle 3D scenes?

In addition to the camera angle, 3D scenes also need light to be visible. For one project I solved this by creating a testing environment scene which would be instantiated if no Camera3D was detected upon startup. The scene included basic lighting, orbiting camera and floor for physics objects. In the end such solution would need to be tailored for every project anyway, so I'm not sure if there's anything we can make in the core.

**Comment by manuq** [comment]
I have been affected by the default behavior many times, and I've seen newcomers frustated by it too. I think the problem is with scenes that are meant to be instantiated from other scenes, so (as @alexgamadev said) they are built around the center of coordinates. So if you want to test the scene in isolation it will show only a quarter (the bottom-right) of the scene. Having (0, 0) as the top-left corner is what you want for main scenes (if they don't have a camera).

Now, I think that the **Project Camera Override** tool is usually what the user expects: to run the scene more-or-less how it looks in the editor viewport. The problem is that it has to be turned on every time, going back to the editor. Besides, this tool deserves more documentation. So what I would like is to turn the camera override on by default if the scene a) is not the main scene b) doesn't have a camera already and c) is not a Control scene.

I have created a plugin that does this as a proof of concept: https://github.com/manuq/godot_addon_play_centered

I think #7213 is related, because it will be more noticeable running the scene in place of the editor viewport. A newcomer would expect to see things as in the viewport if no camera is set.

**Comment by ClarkBrendan** [comment]
Did anything actually come of this? I've been having this same problem a lot. It really bothers me that for most of my scenes everything is centered on the origin to make it easy to import them into other scenes, but then when I try to test a scene separately, I have to override the camera every time to actually see 3/4 of the scene. I was positive this would be an option in the Window project settings where it lets you resize the window/viewport, but I can't find any option to do so. Couldn't there just be two variables for the viewport offset that default to 0 but can be changed to move around the viewport? There could even be a toggle to set it to be centered which would just set the offsets to: -width/2 and -height/2. I don't have any experience with godot's program under the hood, but this can't be a difficult change, can it?
